openmc/mgxs/library.py

```diff
@@ -1072,32 +1087,80 @@ def create_mg_library(self, xs_type='macro', xsdata_names=None,
         # Initialize file
         mgxs_file = openmc.MGXSLibrary(self.energy_groups)
 
-        # Create the xsdata object and add it to the mgxs_file
-        for i, domain in enumerate(self.domains):
-            if self.by_nuclide:
-                nuclides = domain.get_nuclides()
+        # Get the number of domains to size arrays with
+        if self.domain_type is 'mesh':
+            num_domains = np.sum(d.num_mesh_cells for d in self.domains)
+        else:
+            num_domains = len(self.domains)
+
+        # Set id names
+        if xs_ids is not None:
+            if isinstance(xs_ids, basestring):
+                # If we only have a string lets convert it now to a list
+                # of strings.
+                all_xs_ids = [xs_ids] * num_domains
             else:
-                nuclides = ['total']
-            for nuclide in nuclides:
-                # Build & add metadata to XSdata object
-                if xsdata_names is None:
-                    xsdata_name = 'set' + str(i + 1)
-                else:
-                    xsdata_name = xsdata_names[i]
-                if nuclide is not 'total':
-                    xsdata_name += '_' + nuclide
+                cv.check_iterable_type('xs_ids', xs_ids, basestring)
+                cv.check_length('xs_ids', xs_ids, num_domains, num_domains)
+                all_xs_ids = xs_ids
 
-                xsdata = self.get_xsdata(domain, xsdata_name, nuclide=nuclide,
-                                         xs_type=xs_type, xs_id=xs_ids[i],
-                                         tabular_legendre=tabular_legendre,
-                                         tabular_points=tabular_points)
+        else:
+            all_xs_ids = ['1m'] * num_domains
 
-                mgxs_file.add_xsdata(xsdata)
+        if self.domain_type == 'mesh':
+            # Create the xsdata objects and add to the mgxs_file
+            i = 0
+            for domain in self.domains:
+                if self.by_nuclide:
+                    raise NotImplementedError("Mesh domains do not currently "
+                                              "support nuclidic tallies")
+                for subdomain in domain.cell_generator():
+                    # Build & add metadata to XSdata object
+                    if xsdata_names is None:
+                        xsdata_name = 'set' + str(i + 1)
+                    else:
+                        xsdata_name = xsdata_names[i]
+
+                    # Create XSdata and Macroscopic for this domain
+                    xsdata = self.get_xsdata(domain, xsdata_name,
```

Comment:
Are you using all of the keyword arguments supported by `get_xsdata(...)`? If so, can you remove the keywords here to compress this code  a bit?